lib/Dialect/Seq/SeqOps.cpp

```diff
@@ -675,20 +684,25 @@ OpFoldResult FirRegOp::fold(FoldAdaptor adaptor) {
   if (!isTrivialFeedback && !isNeverClocked)
     return {};
 
-  // If the register has a const reset value, we can replace it with that.
-  // We cannot replace it with a non-constant reset value.
-  if (auto resetValue = getResetValue()) {
-    if (auto *op = resetValue.getDefiningOp())
-      if (op->hasTrait<OpTrait::ConstantLike>())
-        return resetValue;
-    return {};
-  }
+  // If the register has a const reset value, and no preset, we can replace it
+  // with the const reset. We cannot replace it with a non-constant reset value.
+  if (!presetAttr)
+    if (auto resetValue = getResetValue()) {
+      if (auto *op = resetValue.getDefiningOp())
+        if (op->hasTrait<OpTrait::ConstantLike>())
+          return resetValue;
+      return {};
+    }
 
   // Otherwise we want to replace the register with a constant 0. For now this
   // only works with integer types.
   auto intType = dyn_cast<IntegerType>(getType());
   if (!intType)
     return {};
+  // If a non-zero preset present, then we cannot replace with zero.
+  if (presetAttr)
+    if (!presetAttr.getValue().isZero())
+      return {};
```

Comment:
I think we can replace it with `presetAttr` even when preset is not zero?

Reply:
You are right, addressed it.